packages/hyperion-autologging/src/ALSurfaceProxy.ts

```diff
@@ -40,34 +43,61 @@ import * as IReactDOM from "@hyperion/hyperion-react/src/IReactDOM";
 //   return props.children;
 // }
 
-export type InitOptions =
+export type InitOptions<
+  DataType extends ALFlowletDataType,
+  FlowletType extends Flowlet<DataType>,
+  FlowletManagerType extends FlowletManager<FlowletType>> =
   Readonly<{
     ReactModule: { createElement: typeof React.createElement };
     IReactDOMModule: IReactDOM.IReactDOMModuleExports;
-    domSurfaceAttributeName?: string;
+    surfaceComponent?: SurfaceComponent<DataType, FlowletType, FlowletManagerType>;
+    flowletManager: FlowletManagerType;
   }>;
 
-export function init(options: InitOptions): void {
-  const { domSurfaceAttributeName = AUTO_LOGGING_SURFACE, IReactDOMModule } = options;
-
+export function init<
+  DataType extends ALFlowletDataType,
+  FlowletType extends Flowlet<DataType>,
+  FlowletManagerType extends FlowletManager<FlowletType>>(options: InitOptions<DataType, FlowletType, FlowletManagerType>): void {
+  const { IReactDOMModule, ReactModule, flowletManager } = options;
   /**
    * When createPortal is called, the react components will be added to a
    * separate container DOM node and shown in place later.
    * Although DOM tree hierarchy is broken, the React Context hierarchy will
-   * continue to work. So, we use that fact and assign the right surface attribute
-   * to the container node. This way, when we walk up the DOM tree, we find the
-   * right surface value.
+   * continue to work. So, we use that fact and wrap the node in another Surface
+   * with the same flowlet and surface from the original surface context.
    */
   IReactDOMModule.createPortal.onArgsMapperAdd(args => {
-    const [_node, container] = args;
-    if (container instanceof HTMLElement) {
-      // args[0] = options.ReactModule.createElement(SurfaceProxy, { container }, node);
-      const { surface } = useALSurfaceContext();
+    const [node, _container] = args;
+
+    if (node != null && React.isValidElement(node)) {
+      const { surface, flowlet: surfaceFlowlet } = useALSurfaceContext<DataType, FlowletType>();
       if (surface != null) {
-        container.setAttribute(domSurfaceAttributeName, surface);
+        // If the surface of the flowlet is the same surface as the one we are proxy wrapping
+        // because of portal, pass that to this wrapping surface component.
+        let flowlet = surfaceFlowlet;
+        if (flowlet == null) {
+          const topFlowlet = flowletManager.top();
+          if (topFlowlet == null || topFlowlet?.name !== surface) {
+            flowlet = flowletManager.push(new flowletManager.flowletCtor(surface, topFlowlet));
+          } else {
+            flowlet = topFlowlet;
+          }
+        }
+        if (options.surfaceComponent) {
```

Comment:
why should it be optional?

Reply:
`Surface.init` would require it then,  which is not possible to provide as is because it's nested within.   Is there a way to have it's init options (`SurfaceProxy`) required except for `surfaceComponent` and only require it via `SurfaceProxy.init`?  Would that warrant two separate types,  or is there a trick to require it only when `SurfaceProxy.init` is called but still share the init requirements for when Surface.init is called?

Reply:
Updated in latest push to use two types,  but if there's a better way I can change it